internal/internal_versioning_client.go

```diff
@@ -0,0 +1,319 @@
+		// Build ID is the one mentioned in the first unconditional Assignment Rule.
+		// If there is no default Build ID, the result for the
+		// unversioned queue will be returned.
+		Versions *TaskQueueVersionSelection
+		// Task queue types to report info about. If not specified, all types are considered.
+		TaskQueueTypes []TaskQueueType
+		// Include list of pollers for requested task queue types and versions.
+		ReportPollers bool
+		// Include task reachability for the requested versions and all task types
+		// (task reachability is not reported per task type).
+		ReportTaskReachability bool
+	}
+
+	// WorkerVersionCapabilities includes a worker's build identifier
+	// and whether it is choosing to use the versioning feature.
+	// It is an optional component of [PollerInfo].
+	WorkerVersionCapabilities struct {
+		BuildID       string
+		UseVersioning bool
+	}
+
+	// PollerInfo provides information about a worker/client polling a task queue.
+	// It is used by [TaskQueueTypeInfo].
+	PollerInfo struct {
+		// Time of the last poll. A value of zero means it was not set.
+		LastAccessTime time.Time
+		// The identity of the worker/client who is polling this task queue.
+		Identity string
+		// Polling rate. A value of zero means it was not set.
+		RatePerSecond float64
+		// Optional poller versioning capabilities. Available when a worker has opted into the worker versioning feature.
+		WorkerVersionCapabilities *WorkerVersionCapabilities
+	}
+
+	// TaskQueueTypeInfo specifies task queue information per task type and Build ID.
+	// It is included in [TaskQueueVersionInfo].
+	TaskQueueTypeInfo struct {
+		// Poller details for this task queue category.
+		Pollers []PollerInfo
+	}
+
+	// TaskQueueVersionInfo includes task queue information per Build ID.
+	// It is part of [TaskQueueDescription].
+	TaskQueueVersionInfo struct {
+		// Task queue info per task type.
+		TypesInfo map[TaskQueueType]TaskQueueTypeInfo
+		// The category of tasks that may reach a versioned worker of a certain Build ID.
+		TaskReachability BuildIDTaskReachability
+	}
+
+	// TaskQueueDescription is the response to [Client.DescribeTaskQueueEnhanced].
+	TaskQueueDescription struct {
+		// Task queue information for each Build ID. Empty string as key value means unversioned.
+		VersionsInfo map[string]TaskQueueVersionInfo
+	}
+)
+
+func (o *DescribeTaskQueueEnhancedOptions) validateAndConvertToProto(namespace string) (*workflowservice.DescribeTaskQueueRequest, error) {
+	if namespace == "" {
+		return nil, errors.New("missing namespace argument")
+	}
+
+	if o.TaskQueue == "" {
+		return nil, errors.New("missing task queue field")
+	}
+
+	taskQueueTypes := make([]enumspb.TaskQueueType, len(o.TaskQueueTypes))
+	for i, t := range o.TaskQueueTypes {
+		taskQueueTypes[i] = taskQueueTypeToProto(t)
+	}
+
+	opt := &workflowservice.DescribeTaskQueueRequest{
+		Namespace: namespace,
+		TaskQueue: &taskqueuepb.TaskQueue{
+			// Sticky queues not supported
```

Comment:
nit: Think we should note this in `DescribeTaskQueueEnhanced`?

Reply:
It is in the (internal) `DescribeTaskQueueEnhancedOptions` docs, but yes, it is worth making this more clear at the top level.